fossunited/foss_hackathon/doctype/foss_hackathon_localhost/foss_hackathon_localhost.py

```diff
@@ -35,55 +35,70 @@ class FOSSHackathonLocalHost(Document):
     pass
 
     def before_insert(self):
-        self.assign_localhost_organizer_role()
+        members = [member.profile for member in self.organizers]
+        self.assign_localhost_organizer_role(members)
 
-    def on_update(self):
-        self.check_if_member_removed()
+    def on_trash(self):
+        members = [member.profile for member in self.organizers]
+        self.remove_organizer_role(members)
 
     def before_save(self):
-        if self.has_value_changed("organizers") and (
-            len(self.organizers) > len(self.get_doc_before_save().organizers)
-        ):
-            self.assign_localhost_organizer_role()
+        self.handle_roles()
 
-    def assign_localhost_organizer_role(self):
-        for member in self.organizers:
+    def handle_roles(self):
+        if self.is_new():
+            return
+
+        organizers = self.get("organizers")
+        organizers_old = self.get_doc_before_save().get("organizers")
+
+        # get members that are in curr member but not in old members
+        new_members = [
+            member.profile
+            for member in organizers
+            if member.profile not in [m.profile for m in organizers_old]
+        ]
+        removed_members = [
+            member.profile
+            for member in organizers_old
+            if member.profile not in [m.profile for m in organizers]
+        ]
```

Comment:
can you think about the time-complexity of this code? we're unnecessarily iterating over loops. how about

```
# There is no point of duplicates here so we could use sets instead of lists
new_profiles = {member.profile for member in organizers}
old_profiles = {member.profile for member in organizers_old}

# using sets makes the comparisons faster
profiles_to_add = list(new_profiles - old_profiles)
profiles_to_remove = list(ld_profiles - new_profiles)
```

please correct me if I fundamentally misunderstood the code

Reply:
This works, changing the code